urt/man/man3/buildmap.3
.\" Copyright (c) 1986, University of Utah
.TH BUILDMAP 3 3/6/85 3
.UC 4
.SH NAME
buildmap \- create a color map array from an RLE file header.
.SH SYNOPSIS
.B
#include <rle.h>
.na
.HP
.B
rle_pixel ** buildmap( the_hdr, minmap, orig_gamma, new_gamma )
.PD 0
.PP
.B
rle_hdr * the_hdr;
.br
.B
int minmap;
.br
.B 
double orig_gamma, new_gamma;
.PD
.SH DESCRIPTION
The color map in the 
.IR rle_hdr (3)
structure is not in the most easily used form.  The function
.I buildmap
returns a pointer to a colormap array with certain minimum dimensions,
making it a little easier to implement color mapping in a program.
The color map from first argument,
.IR the_hdr ,
is used to build the result.  If no map is present in
.IR the_hdr ,
then an identity map of the minimum size will be returned.
.PP
The returned color map will have at least
.I minmap
rows or channels, each of which is at least 256 entries long (so that
indexing into the color map with an 8 bit 
.I rle_pixel
value will always succeed.)

The color map from 
.IR the_hdr
will be composed with a gamma compensation curve to account for the
gamma of the display for which the input color map was presumably computed.
The argument
.I orig_gamma
specifies the gamma of the compensation curve.  It would typically be
the gamma of the original display.

If 
.I gamma
is 0, then if a picture comment 
.I image_gamma=i_gamma
is present, \fIgamma\fP will be set to \fI1.0/i_gamma\fP.
Otherwise, if a comment
.I display_gamma=d_gamma
is present, \fIgamma\fP will be set to \fId_gamma\fP.  The gamma
compensation value for pixel \fIi\fP is \fI255*(i/255)^gamma\fP.

If this color map will be used directly for another display, the gamma
of this new display should be passed in
.IR new_gamma .

The returned value is a pointer to an array of pointers to arrays of 
.I rle_pixel
values.  It may be doubly indexed in C code, so that if
.I cmap
is the return value, the RGB color mapping for a pixel
.I pixval
is
.IR "(cmap[0][pixval], cmap[1][pixval], cmap[2][pixval])" .
.SH NOTES
Generally, unless the user explicitly specifies the image or original
display gamma (e.g., as with the \fB\-i\fP or \fB\-I\fP flags of
.IR getx11 (1),
you should pass 0 for
.IR orig_gamma .
This lets 
.I buildmap
use the value from \fIthe_hdr\fP, if it is present.

If you are going to use the result of buildmap to generate values to
be dithered,
.I new_gamma
should always be 1.0, and the display gamma (\fB-g\fP in
.IR getx11 )
should be passed to
.IR dithermap (3).  
If you are not planning to dither, then pass the
user supplied display gamma as 
.IR new_gamma .
.SH SEE ALSO
.IR dithermap (3),
.IR rle_hdr (3),
.IR librle (3),
.IR RLE (5).
.SH AUTHOR
Spencer W. Thomas,
University of Utah



urt/lib/buildmap.c
/*
 * This software is copyrighted as noted below.  It may be freely copied,
 * modified, and redistributed, provided that the copyright notice is 
 * preserved on all copies.
 * 
 * There is no warranty or other guarantee of fitness for this software,
 * it is provided solely "as is".  Bug reports or fixes may be sent
 * to the author, who may or may not act on them as he desires.
 *
 * You may not include this software in a program or other software product
 * without supplying the source, or without informing the end-user that the 
 * source is available for no extra charge.
 *
 * If you modify this software, you should include a notice giving the
 * name of the person performing the modification, the date of modification,
 * and the reason for such modification.
 */
/* 
 * buildmap.c - Build a color map from the RLE file color map.
 * 
 * Author:	Spencer W. Thomas
 * 		Computer Science Dept.
 * 		University of Utah
 * Date:	Sat Jan 24 1987
 * Copyright (c) 1987, University of Utah
 */

#include <stdio.h>
#include <rle.h>
#include <math.h>

#ifdef VOID_STAR
void *malloc();
#else
char *malloc();
#endif
void free();

/*****************************************************************
 * TAG( buildmap )
 * 
 * Returns a color map that can easily be used to map the pixel values in
 * an RLE file.  Map is built from the color map in the input file.
 * Inputs:
 * 	the_hdr:	rle_hdr structure containing color map.
 *	minmap:		Minimum number of channels in output map.
 *	orig_gamma:	Adjust color map for this image gamma value
 *			(1.0 means no adjustment).
 *	new_gamma:	Gamma of new display.
 * Outputs:
 * 	Returns an array of pointers to arrays of rle_pixels.  The array
 *	of pointers contains max(ncolors, ncmap) elements, each 
 *	array of pixels contains 2^cmaplen elements.  The pixel arrays
 *	should be considered read-only.
 * Assumptions:
 * 	[None]
 * Algorithm:
 *	Ensure that there are at least ncolors rows in the map, and
 *	that each has at least 256 elements in it (largest map that can
 *	be addressed by an rle_pixel).
 */
rle_pixel **
buildmap( the_hdr, minmap, orig_gamma, new_gamma )
rle_hdr *the_hdr;
int minmap;
double orig_gamma;
double new_gamma;
{
    rle_pixel ** cmap, * gammap;
    double gamma;
    register int i, j;
    int maplen, cmaplen, ncmap, nmap;

    if ( the_hdr->ncmap == 0 )	/* make identity map */
    {
	nmap = (minmap < the_hdr->ncolors) ? the_hdr->ncolors : minmap;
	cmap = (rle_pixel **)malloc( nmap * sizeof(rle_pixel *) );
	cmap[0] = (rle_pixel *)malloc( 256 * sizeof(rle_pixel) );
	for ( i = 0; i < 256; i++ )
	    cmap[0][i] = i;
	for ( i = 1; i < nmap; i++ )
	    cmap[i] = cmap[0];
	maplen = 256;
	ncmap = 1;		/* number of unique rows */
    }
    else			/* make map from the_hdr */
    {
	/* Map is at least 256 long */
	cmaplen = (1 << the_hdr->cmaplen);
	if ( cmaplen < 256 )
	    maplen = 256;
	else
	    maplen = cmaplen;

	if ( the_hdr->ncmap == 1 )	/* make "b&w" map */
	{
	    nmap = (minmap < the_hdr->ncolors) ?
		the_hdr->ncolors : minmap;
	    cmap = (rle_pixel **)malloc( nmap * sizeof(rle_pixel *) );
	    cmap[0] = (rle_pixel *)malloc( maplen * sizeof(rle_pixel) );
	    for ( i = 0; i < maplen; i++ )
		if ( i < cmaplen )
		    cmap[0][i] = the_hdr->cmap[i] >> 8;
		else
		    cmap[0][i] = i;
	    for ( i = 1; i < nmap; i++ )
		cmap[i] = cmap[0];
	    ncmap = 1;
	}
	else if ( the_hdr->ncolors <= the_hdr->ncmap )
	{
	    nmap = (minmap < the_hdr->ncmap) ? the_hdr->ncmap : minmap;
	    cmap = (rle_pixel **)malloc( nmap * sizeof(rle_pixel *) );
	    for ( j = 0; j < the_hdr->ncmap; j++ )
	    {
		cmap[j] = (rle_pixel *)malloc( maplen * sizeof(rle_pixel) );
		for ( i = 0; i < maplen; i++ )
		    if ( i < cmaplen )
			cmap[j][i] = the_hdr->cmap[j*cmaplen + i] >> 8;
		    else
			cmap[j][i] = i;
	    }
	    for ( i = j, j--; i < nmap; i++ )
		cmap[i] = cmap[j];
	    ncmap = the_hdr->ncmap;
	}
	else			/* ncolors > ncmap */
	{
	    nmap = (minmap < the_hdr->ncolors) ?
		the_hdr->ncolors : minmap;
	    cmap = (rle_pixel **)malloc( nmap * sizeof(rle_pixel *) );
	    for ( j = 0; j < the_hdr->ncmap; j++ )
	    {
		cmap[j] = (rle_pixel *)malloc( maplen * sizeof(rle_pixel) );
		for ( i = 0; i < maplen; i++ )
		    if ( i < cmaplen )
			cmap[j][i] = the_hdr->cmap[j*cmaplen + i] >> 8;
		    else
			cmap[j][i] = i;
	    }
	    for( i = j, j--; i < nmap; i++ )
		cmap[i] = cmap[j];
	    ncmap = the_hdr->ncmap;
	}
    }
	    
    /* Gamma compensate if requested */
    if ( orig_gamma == 0 )
    {
	char *v;
	if ( (v = rle_getcom( "image_gamma", the_hdr )) != NULL )
	{
	    orig_gamma = atof( v );
	    /* Protect against bogus information */
	    if ( orig_gamma == 0.0 )
		orig_gamma = 1.0;
	    else
		orig_gamma = 1.0 / orig_gamma;
	}
	else if ( (v = rle_getcom( "display_gamma", the_hdr )) != NULL)
	{
	    orig_gamma = atof( v );
	    /* Protect */
	    if ( orig_gamma == 0.0 )
		orig_gamma = 1.0;
	}
	else
	    orig_gamma = 1.0;
    }
    /* Now, compensate for the gamma of the new display, too. */
    if ( new_gamma != 0.0 )
	gamma = orig_gamma / new_gamma;
    else
	gamma = orig_gamma;

    if ( gamma != 1.0 )
    {
	gammap = (rle_pixel *)malloc( 256 * sizeof(rle_pixel) );
	for ( i = 0; i < 256; i++ )
	    gammap[i] = (int)(0.5 + 255.0 * pow( i / 255.0, gamma ));
	for ( i = 0; i < ncmap; i++ )
	    for ( j = 0; j < maplen; j++ )
		cmap[i][j] = gammap[cmap[i][j]];
	free( gammap );
    }

    return cmap;
}
